File: native/android/app/src/main/java/org/membercovenantpath/viewer/model/Member.kt
package org.membercovenantpath.viewer.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

/**
 * One tracked person, mirroring the `members` table columns the Flutter dashboard selects
 * (see SPEC.md "Select string" / dashboard_page.dart `_columns`). RLS returns only rows the
 * signed-in user may see — the app does NO filtering of its own.
 *
 * Yes/No status fields are kept as raw strings ("Yes" / "No" / "N/A" / sentinel) because the
 * milestone logic compares against the literal "Yes" exactly like the Flutter app does, and
 * because sentinels ("needs-profile-api", "blocked: ...") mean "unknown", not "No".
 *
 * `@SerialName` values match the snake_case DB columns 1:1.
 */
@Serializable
data class Member(
    @SerialName("person_uuid") val personUuid: String? = null,
    @SerialName("stake_id") val stakeId: String? = null,
    @SerialName("unit_id") val unitId: String? = null,
    @SerialName("name") val name: String? = null,
    @SerialName("unit_name") val unitName: String? = null,
    @SerialName("kind") val kind: String? = null, // 'new_member' | 'investigator' | 'returning'
    @SerialName("baptism_date") val baptismDate: String? = null,
    @SerialName("baptism_goal_date") val baptismGoalDate: String? = null,
    @SerialName("birth_date") val birthDate: String? = null,
    @SerialName("membership_duration") val membershipDuration: String? = null,
    @SerialName("sex") val sex: String? = null, // 'M' / 'F'

    // Yes/No status fields.
    @SerialName("friends") val friends: String? = null,
    @SerialName("friends_count") val friendsCount: Int? = null,
    @SerialName("aaronic_priesthood") val aaronicPriesthood: String? = null,
    @SerialName("melchizedek_priesthood") val melchizedekPriesthood: String? = null,
    @SerialName("calling") val calling: String? = null,
    @SerialName("ministering_brothers_sisters") val ministeringBrothersSisters: String? = null,
    @SerialName("ministering_assignment") val ministeringAssignment: String? = null,
    @SerialName("temple_recommend") val templeRecommend: String? = null, // Active / Expired / No
    @SerialName("patriarchal_blessing") val patriarchalBlessing: String? = null,
    @SerialName("living_ordinance") val livingOrdinance: String? = null,

    // Rich detail subtree for the person-detail page (may be partial / null).
    @SerialName("details") val details: JsonObject? = null,
    @SerialName("photo_url") val photoUrl: String? = null,
) {
    val isInvestigator: Boolean get() = kind == "investigator"

    /** Raw value for a milestone "yes/no" key, by the same DB column name the milestone uses. */
    fun status(field: String): String? = when (field) {
        "friends" -> friends
        "aaronic_priesthood" -> aaronicPriesthood
        "melchizedek_priesthood" -> melchizedekPriesthood
        "calling" -> calling
        "ministering_brothers_sisters" -> ministeringBrothersSisters
        "ministering_assignment" -> ministeringAssignment
        "temple_recommend" -> templeRecommend
        "patriarchal_blessing" -> patriarchalBlessing
        "living_ordinance" -> livingOrdinance
        else -> null
    }
}

File: apps/web/src/lib/member.ts
// The member shape the dashboard reads from Supabase. Mirrors `_columns` in
// apps/viewer/lib/dashboard_page.dart — the single source for the Supabase select. Fields are
// loosely typed (Supabase returns JSON) and accessed via helpers; `details` is the rich one-work
// subtree used by the person-detail page.

/** Exact column list selected from `members` (mirrors the Flutter `_columns`). */
export const MEMBER_COLUMNS =
  'person_uuid, stake_id, unit_id, name, unit_name, baptism_date, birth_date, membership_duration, sex, friends, friends_count, ' +
  'aaronic_priesthood, melchizedek_priesthood, calling, ministering_brothers_sisters, ' +
  'ministering_assignment, temple_recommend, patriarchal_blessing, living_ordinance, ' +
  'family_name_prepared, first_temple_visit, details, photo_url, ' +
  'kind, baptism_goal_date, field_meta';

/** A loosely-typed member record (mirrors the Dart `Map<String, dynamic>`). */
export type Member = Record<string, unknown>;

export function str(m: Member, key: string): string {
  const v = m[key];
  return v == null ? '' : String(v);
}

// --- sync sentinels (mirror covenant_path/report.py + backend/db.py _SENTINELS) -----------------
// The backend writes these placeholder strings when a field couldn't be fetched this run (so the
// merge-upsert preserves last-good). They are TECHY/internal — only ADMINS should ever see the raw
// value; regular leaders get human-friendly text. (NEVER leak the raw sentinel into a leader's cell.)
export const NEEDS_PROFILE = 'needs-profile-api';
export const BLOCKED_PREFIX = 'blocked'; // 'blocked: insufficient calling access'

/** True when a value is one of the backend's internal sentinels (not real member data). */
export function isSentinel(v: unknown): boolean {
  const s = String(v ?? '');
  return s === NEEDS_PROFILE || s.startsWith(BLOCKED_PREFIX);
}

/**
 * Map a possibly-sentinel field value to what should be DISPLAYED, audience-aware:
 *  - admins see the raw sentinel verbatim (so they can diagnose "needs-profile-api" / "blocked: …");
 *  - everyone else sees friendly text — "Not available yet" for a pending/blocked field, never the
 *    raw techy string. A real value (Yes/No/Active/a date) is returned unchanged for both.
 * `blank` is what a non-admin sees for a sentinel ('—' in dense tables, 'Not available yet' in detail).
 */
export function displayFieldValue(
  v: unknown,
  isAdmin: boolean,
  blank: '—' | 'Not available yet' = 'Not available yet',
): string {
  const s = String(v ?? '');
  if (isSentinel(s)) return isAdmin ? s : blank;
  return s;
}

export function numOrNull(v: unknown): number | null {
  if (v == null) return null;
  const n = typeof v === 'number' ? v : Number(v);
  return Number.isFinite(n) ? n : null;
}

/** Stable identity for de-duping people in stat buckets (person_uuid → name → fallback). */
export function memberId(m: Member): string {
  const u = m['person_uuid'];
  if (u != null && String(u).length > 0) return String(u);
  const n = m['name'];
  if (n != null && String(n).length > 0) return String(n);
  return JSON.stringify(m);
}

export function isInvestigator(m: Member): boolean {
  return m['kind'] === 'investigator';
}

/** The rich one-work detail subtree, or null for rows synced before the schema change. */
export function detailsOf(m: Member): Record<string, unknown> | null {
  const d = m['details'];
  return d && typeof d === 'object' && !Array.isArray(d) ? (d as Record<string, unknown>) : null;
}

/** Per-field freshness map ({field:{f:last-fetched, t:last-tried}}, migration 0036), or {}. */
export function fieldMetaOf(m: Member): Record<string, { f?: string; t?: string }> {
  const fm = m['field_meta'];
  return fm && typeof fm === 'object' && !Array.isArray(fm)
    ? (fm as Record<string, { f?: string; t?: string }>)
    : {};
}

export type Freshness = 'fresh' | 'warn' | 'error' | 'never' | 'unknown';

/** Staleness of a field from its last-SUCCESSFUL fetch: >3d warn, >7d error, no fetch = never. */
export function freshness(m: Member, field: string): { state: Freshness; fetched?: string } {
  const meta = fieldMetaOf(m)[field];
  if (!meta) return { state: 'unknown' };
  if (!meta.f) return { state: 'never' };
  const days = (Date.now() - new Date(meta.f).getTime()) / 86_400_000;
  if (Number.isNaN(days)) return { state: 'unknown', fetched: meta.f };
  if (days > 7) return { state: 'error', fetched: meta.f };
  if (days > 3) return { state: 'warn', fetched: meta.f };
  return { state: 'fresh', fetched: meta.f };
}
